### PR TITLE
feat(api): Decision Notification Listener for feature variable APIs.

### PR DESCRIPTION
**Pull request by mfahadahmed**
## Summary
- Implemented onDecisionListener in activate and getVariation APIs.

## Test plan
- Added unit tests.

### CODE REVIEW

File: packages/optimizely-sdk/lib/core/decision_service/index.tests.js [hunk 1]
```diff
@@ -729,22 +729,22 @@ describe('lib/core/decision_service', function() {
                     'variables': [
                       {
                         'id': '4792309476491264',
-                        'value': '10'
+                        'value': '40'
```
Comment (aliabbasrizvi) [inline packages/optimizely-sdk/lib/core/decision_service/index.tests.js]: Why the change in these values?

File: packages/optimizely-sdk/lib/optimizely/index.tests.js [hunk 2]
```diff
@@ -2519,6 +2521,504 @@ describe('lib/optimizely', function() {
...
+                });
+              });
+            });
+
+            describe('not bucketed into an experiment or a rollout', function() {
+              beforeEach(function() {
+                sandbox.stub(optlyInstance.decisionService, 'getVariationForFeature').returns({
+                  experiment: null,
+                  variation: null,
+                  decisionSource: null,
```
Comment (aliabbasrizvi) [inline packages/optimizely-sdk/lib/optimizely/index.tests.js]: I feel the actual `decisionService` needs to be updated to set decisionSource as rollout.

Thoughts @jordangarcia @mjc1283 ?

File: packages/optimizely-sdk/lib/core/decision_service/index.tests.js (hunk 1 above)
Comment (mfahadahmed) [reply packages/optimizely-sdk/lib/core/decision_service/index.tests.js]: I used this variation and its related variables to verify default values when feature toggle is off. But now I realized that this can easily be done without changing values in the datafile. 

Reverted decision_service and testdata files. Thanks for pointing this out.

File: packages/optimizely-sdk/lib/optimizely/index.js [hunk 3]
```diff
@@ -563,17 +564,51 @@ Optimizely.prototype._getFeatureVariableForType = function(featureKey, variableK
...
+  var featureEnabled = false;
+  var variableValue = variable.defaultValue;
   var decision = this.decisionService.getVariationForFeature(featureFlag, userId, attributes);
-  var variableValue;
+  
   if (decision.variation !== null) {
-    variableValue = projectConfig.getVariableValueForVariation(this.configObj, variable, decision.variation, this.logger);
-    this.logger.log(LOG_LEVEL.INFO, sprintf(LOG_MESSAGES.USER_RECEIVED_VARIABLE_VALUE, MODULE_NAME, variableKey, featureFlag.key, variableValue, userId));
+    featureEnabled = decision.variation.featureEnabled;
+    if (featureEnabled === true) {
```
Comment (jordangarcia) [inline packages/optimizely-sdk/lib/optimizely/index.js]: This logic is incorrect.  

Feature variables should always be the variation variable value, even if the feature is disabled.

File: packages/optimizely-sdk/lib/optimizely/index.js [hunk 4]
```diff
@@ -563,17 +564,51 @@ Optimizely.prototype._getFeatureVariableForType = function(featureKey, variableK
...
+      variableValue = projectConfig.getVariableValueForVariation(this.configObj, variable, decision.variation, this.logger);
+      this.logger.log(LOG_LEVEL.INFO, sprintf(LOG_MESSAGES.USER_RECEIVED_VARIABLE_VALUE, MODULE_NAME, variableKey, featureFlag.key, variableValue, userId));
+    } else {
+      this.logger.log(LOG_LEVEL.INFO, sprintf(LOG_MESSAGES.FEATURE_NOT_ENABLED_RETURN_DEFAULT_VARIABLE_VALUE, MODULE_NAME, featureFlag.key, userId, variableKey));
+      featureEnabled = false;
```
Comment (jordangarcia) [inline packages/optimizely-sdk/lib/optimizely/index.js]: There is no reason to re-card `featureEnabled` here.

File: packages/optimizely-sdk/lib/optimizely/index.tests.js [hunk 5]
```diff
@@ -2519,6 +2521,511 @@ describe('lib/optimizely', function() {
...
+        });
+
+        describe('feature management', function() {
+          var sandbox = sinon.sandbox.create();
+
+          beforeEach(function() {
+            optlyInstance = new Optimizely({
+              clientEngine: 'node-sdk',
+              datafile: testData.getTestProjectConfigWithFeatures(),
+              eventBuilder: eventBuilder,
```
Comment (jordangarcia) [inline packages/optimizely-sdk/lib/optimizely/index.tests.js]: `eventBuilder` isn't a valid option to pass to `new Optimizely`

File: packages/optimizely-sdk/lib/optimizely/index.tests.js [hunk 6]
```diff
@@ -2519,6 +2521,511 @@ describe('lib/optimizely', function() {
...
+                      sourceVariationKey: 'variation'
+                    }
+                  });
+                });
+              });
+      
+              describe('when the variation is toggled OFF', function() {
+                beforeEach(function() {
+                  // Setting featureEnabled to false for verifying feature toggle off.
+                  optlyInstance.configObj.experimentKeyMap.testing_my_feature.variations[0].featureEnabled = false;
```
Comment (jordangarcia) [inline packages/optimizely-sdk/lib/optimizely/index.tests.js]: We shouldn't be reaching into the `configObj` like this.

We should change the test datafile that were using to include a feature where a variation has `featureEnabled = false`

File: packages/optimizely-sdk/lib/optimizely/index.tests.js [hunk 7]
```diff
@@ -2519,6 +2521,511 @@ describe('lib/optimizely', function() {
...
+                  var experiment = optlyInstance.configObj.experimentKeyMap.testing_my_feature;
```
Comment (jordangarcia) [inline packages/optimizely-sdk/lib/optimizely/index.tests.js]: use `projectConfig.getExperimentFromKey(optlyInstance.configObj, 'testing_my_feature')` instead of directly inspecting the `configObj`

File: packages/optimizely-sdk/lib/optimizely/index.tests.js [hunk 8]
```diff
@@ -2519,6 +2521,511 @@ describe('lib/optimizely', function() {
...
+                    }
+                  });
+                });
+              });
+            });
+
+            describe('bucketed into variation of a rollout with variable values', function() {
+              describe('when the variation is toggled ON', function() {
+                beforeEach(function() {
+                  var experiment = optlyInstance.configObj.experimentKeyMap['594031'];
```
Comment (jordangarcia) [inline packages/optimizely-sdk/lib/optimizely/index.tests.js]: use `projectConfig` function here instead.

File: packages/optimizely-sdk/lib/optimizely/index.tests.js (hunk 2 above)
Comment (jordangarcia) [reply packages/optimizely-sdk/lib/optimizely/index.tests.js]: Yeah I think `decisionSource` should be rollout even if the user is not evaluated into the experiment, as it's jkind of the bottom of evaluation chain.

File: packages/optimizely-sdk/lib/optimizely/index.tests.js [hunk 9]
```diff
@@ -2608,6 +3117,7 @@ describe('lib/optimizely', function() {
       sandbox.stub(uuid, 'v4').returns('a68cf1ad-0393-4e18-af87-efe8f01a7c9c');
       sandbox.stub(fns, 'currentTimestamp').returns(1509489766569);
       clock = sinon.useFakeTimers(new Date().getTime());
+      onDecisionListener = sinon.spy();
```
Comment (jordangarcia) [inline packages/optimizely-sdk/lib/optimizely/index.tests.js]: Is this still needed?

File: packages/optimizely-sdk/lib/optimizely/index.js (hunk 3 above)
Comment (aliabbasrizvi) [reply packages/optimizely-sdk/lib/optimizely/index.js]: @jordangarcia it is actually supposed to be the default feature value

File: packages/optimizely-sdk/lib/optimizely/index.js (hunk 4 above)
Comment (mfahadahmed) [reply packages/optimizely-sdk/lib/optimizely/index.js]: I think this value can be undefined when featureEnabled property is missing from the variation as specified in this test https://github.com/optimizely/javascript-sdk/blob/553e6ffa485a519ed4de399114768ff8bdcc6c5b/packages/optimizely-sdk/lib/optimizely/index.tests.js#L2896

File: packages/optimizely-sdk/lib/optimizely/index.js (hunk 3 above)
Comment (jordangarcia) [reply packages/optimizely-sdk/lib/optimizely/index.js]: @aliabbasrizvi, agree that this is the default behavior.  And the datafile constraint of a feature experiment variation (where `featureEnabled == false`) always has variables that mirror the variable defaults makes it so this works today.

Given the scope of this PR (just adding notificationCenter listener), i'd prefer not to change this core logic. 

Java also mirrors the existing JS implementation https://github.com/optimizely/java-sdk/blob/master/core-api/src/main/java/com/optimizely/ab/Optimizely.java#L615-L668

Comment (mfahadahmed) [reply packages/optimizely-sdk/lib/optimizely/index.js]: Modified the logic to return variation's variable value even when the featureEnabled property is set to false.

File: packages/optimizely-sdk/lib/optimizely/index.js [hunk 10]
```diff
@@ -614,35 +614,46 @@ Optimizely.prototype._getFeatureVariableForType = function(featureKey, variableK
...
+    featureEnabled = decision.variation.featureEnabled;
+    variableValue = projectConfig.getVariableValueForVariation(this.configObj, variable, decision.variation, this.logger);
+    this.logger.log(LOG_LEVEL.INFO, sprintf(LOG_MESSAGES.USER_RECEIVED_VARIABLE_VALUE, MODULE_NAME, variableKey, featureFlag.key, variableValue, userId));
   } else {
-    variableValue = variable.defaultValue;
-    this.logger.log(
-      LOG_LEVEL.INFO,
-      sprintf(LOG_MESSAGES.USER_RECEIVED_DEFAULT_VARIABLE_VALUE, MODULE_NAME, userId, variableKey, featureFlag.key)
-    );
+  	variableValue = variable.defaultValue;
```
Comment (jordangarcia) [inline packages/optimizely-sdk/lib/optimizely/index.js]: fix this indentation.

File: packages/optimizely-sdk/lib/utils/enums/index.js [hunk 11]
```diff
@@ -199,8 +199,8 @@ exports.DECISION_INFO_TYPES = {
  * Optimizely.
  */
 exports.DECISION_SOURCES = {
-  EXPERIMENT: 'experiment',
-  ROLLOUT: 'rollout',
+  EXPERIMENT: 'EXPERIMENT',
```
Comment (jordangarcia) [inline packages/optimizely-sdk/lib/utils/enums/index.js]: This should be lowercase, right?

@aliabbasrizvi

Comment (mfahadahmed) [reply packages/optimizely-sdk/lib/utils/enums/index.js]: I think it will be upper case as @aliabbasrizvi suggested here
https://github.com/optimizely/javascript-sdk/pull/245#discussion_r268360395

**Comment by aliabbasrizvi** [comment]
@jordangarcia can you take a look again and approve/decline?

**Comment by msohailhussain** [comment]
@jordangarcia can you please rereview and let me know if any change is needed?

**Comment by jordangarcia** [comment]
LGTM
